### PR TITLE
Add DCRBaselineProtection and DCROverfittingProtection metrics

### LINKED ISSUE #719: Add `DCROverfittingProtection` metric
### Problem Description
Currently in SDMetrics, we are offering [DisclosureProtection](https://docs.sdv.dev/sdmetrics/metrics/metrics-glossary/disclosureprotection) as the only metric for measuring data privacy. We've vetted this metric to make sure that it's understandable and insightful. However, running this metric requires the user to make some decisions about what to input as known columns and sensitive columns. This means that the privacy score is only ever based on a subset of columns.

We'd like to create a new metric that could be run on an entire set of columns.

### Expected behavior
Add the `DCROverfittingProtection` metric. This metric uses a DCR (distance to closest record) computation to measure whether the synthetic data has been overfit to the real data – as compared to a holdout set.

```python
from sdmetrics.single_table import DCROverfittingProtection

DCROverfittingProtection.compute(
  real_training_data=real_df,
  synthetic_data=synthetic_df,
  real_validation_data=holdout_df,
  metadata=my_metadata)

DCROverfittingProtection.compute_breakdown(
  real_training_data=real_df,
  synthetic_data=synthetic_df,
  real_validation_data=holdout_df,
  metadata=my_metadata)
```

#### `compute_breakdown`
__Args__
- `real_training_data (pd.DataFrame)`: A pandas.DataFrame object containing the real data used for training the synthesizer.
- `synthetic_data (pd.DataFrame)`: A pandas.DataFrame object containing the synthetic data sampled from the synthesizer.
- `real_validation_data (pd.DataFrame)`: A pandas.DataFrame object containing a holdout set of real data. This data should not have been used to train the synthesizer. For this metric to work effectively, this metric should be roughly the same size as the real data.
- `metadata (dict)`: A metadata dictionary that describes the table of data.
- `num_rows_subsample (int, optional)`: The number of synthetic data rows to subsample from the synthetic data. Use this to increase the speed of the computation, if your dataset is large. Default is `None` (do not subsample the data).
- `num_iterations (int, optional)`: The number of iterations to perform when subsampling. The final score will be the average of all iterations. Default is 1.

__Returns__
A dictionary breakdown of the score, with the following information:
- The overall score of the metric
- The percent of synthetic rows that were closer to the training set and the percentage of synthetic rows that were closer to the holdout set.
    - If multiple iterations were used, the percentages should be the average across all iterations.

Example
```python
{
  'score': 0.5,
  'synthetic_data_percentages': {
    'closer_to_training': 0.75,
    'closer_to_holdout': 0.25
  }
}
```

__Computation Algorithm__
1. Preprocessing:
  -> Only consider columns that are statistical (sdtype is numerical, datetime, categorical, or boolean).
  -> Convert all datetimes into UnixTimestamps so they can be treated as numerical values for the rest of the metric
  -> If requested by the user, subsample the synthetic data
2. Go through every row, `s` in the synthetic data and measure the following:
  a. Measure the DCR between `s` and the real data using the algorithm below. Call this distance `d(s, real)`
  b. Measure the DCR of `s`  in the holdout set using the same algorithm. Call this distance `d(s, holdout)`
  -> After Step 1 there will be a `d(s, real)` and `d(s, holdout)` for every row in the synthetic data. 
3. Calculate the percentage of rows where `d(s, real) < d(s, holdout)` (i.e the number of rows that are closer to the real dataset than the holdout dataset). Call this percentage `p`.
4. Return  `min((1-p)*2, 1)`
5. Repeat the above for `num_iterations` requested by the user

_DCR between a row and a dataset_
- Given a row `s` and a dataset `D`, this algorithm computes the distance to the closest record of `D`.
- Compute the record distance between `s` and every single row in `D` using the DCR between two rows algorithm below.
- The DCR is the minimum distance value that we compute between `s` and all rows of `D`.

_DCR between 2 rows (including null values)_
Given a row `s` and a row `d` compute distance between these rows.
- Loop through every column's value (j) in these rows and compute the distance. If these values are called `s_j` and `d_j`, then distance between them is defined based on the type of data:
  - For numerical data, `distance(s_j, d_j) = abs(s_j - d_j) / (range of column)` if both `s_j` and `d_j` are both non-null.
    - The `distance(s_j, d_j) = 0` if both `s_j` and `d_j` are null.
    - The `distance(s_j, d_j) = 1` if only one of `s_j` and `d_j` are null.
  - For categorical data, nulls are considered another category and treated like any other category value.
    - `distance(s_j, d_j) = 0 if s_j == d_j else 1`

The overall distance between rows s and d is the average of the column distances we compute

#### `compute`
The `compute` method should take the same arguments as the `compute_breakdown` method.

The `compute` method should just return the overall `score` parameter calculated by `compute_breakdown`.

### Additional context
See [doc](https://docs.google.com/document/d/1APIEvLfCCptC1bbFgF8bBRLnyAC_3y5pICb_9yCUxuA/edit?tab=t.jkxbunij8p7p#heading=h.va5ntny100zj)

### LINKED ISSUE #720: Add `DCRBaselineProtection` metric
### Problem Description
Currently in SDMetrics, we are offering [DisclosureProtection](https://docs.sdv.dev/sdmetrics/metrics/metrics-glossary/disclosureprotection) as the only metric for measuring data privacy. We've vetted this metric to make sure that it's understandable and insightful. However, running this metric requires the user to make some decisions about what to input as known columns and sensitive columns. This means that the privacy score is only ever based on a subset of columns.

We'd like to create a new metric that could be run on an entire set of columns.

### Expected behavior
Add the `DCRBaselineProtection` metric. This metric uses a DCR (distance to closest record) computation to measure how close the synthetic data is to the real data – as opposed to a baseline of random data. 

```python
from sdmetrics.single_table import DCRBaselineProtection

DCRBaselineProtection.compute(
  real_training_data=real_df,
  synthetic_data=synthetic_df,
  metadata=my_metadata)

DCRBaselineProtection.compute_breakdown(
  real_training_data=real_df,
  synthetic_data=synthetic_df,
  metadata=my_metadata)
```

#### `compute_breakdown`
__Args__
- `real_training_data (pd.DataFrame)`: A pandas.DataFrame object containing the real data used for training the synthesizer.
- `synthetic_data (pd.DataFrame)`: A pandas.DataFrame object containing the synthetic data sampled from the synthesizer.
- `metadata (dict)`: A metadata dictionary that describes the table of data.
- `num_rows_subsample (int, optional)`: The number of synthetic data rows to subsample from the synthetic data. Use this to increase the speed of the computation, if your dataset is large. Default is `None` (do not subsample the data).
- `num_iterations (int, optional)`: The number of iterations to perform when subsampling. The final score will be the average of all iterations. Default is 1.

__Returns__
A dictionary breakdown of the score, with the following information:
- The overall score of the metric
- The median DCR score between the synthetic data and real data, and the median DCR score between the random data and real data.
  - In case of multiple iterations, the median DCR in the breakdown should be the averages across all iterations.

Example
```python
{
  'score': 0.58808,
  'median_DCR_to_real_data': {
    'synthetic_data': 0.12355,
    'random_data_baseline': 0.21009
  }
}
```

__Computation Algorithm__
This metric will use the same DCR measurement as the `DCROverfittingProtection` metric (#719), so this can be abstracted out and shared between the two metrics. It measures the DCR for every synthetic data row vs. the real data. Then it does the same for every random data row vs. the real data. The overall score compares the medians of the two distributions.
- **Note**: We will have to build a method for creating random data for statistical columns (datetime, numerical, categorial, boolean) by learning the ranges. We cannot use SDV or RDT for this, so we should make a helper function that can do this directly in SDMetrics. Want to create random data from the range detected from the real_data. 

1. Preprocessing:
  -> Only consider columns that are statistical (sdtype is numerical, datetime, categorical, or boolean).
  -> Convert all datetimes into UnixTimestamps so they can be treated as numerical values for the rest of the metric
  -> If requested by the user, subsample the synthetic data
2. Go through every row, `s` in the synthetic data and measure the following:
  a. Measure the DCR between `s` and the real data using the algorithm below. Call this distance `d(s, real)`
3. Go through every row, `r` in the random data and measure the following:
  b. Measure the DCR between `r`   in the real set using the same algorithm. Call this distance `d(r, real)`
  -> After Step 2 & 3 there will be a `d(s, real)` and `d(r, real)` for every row in the synthetic data. 
4. Calculate the median value (`synthetic_data_median` for the distribution of values in `d(s, real)`. Also calculate the median value (`random-data_median`) for the distribution of values in `d(r, real)`.
5. The score for this iteration is `min(synthetic_data_media / random_data_median, 1)`
6. Repeat the above for `num_iterations` requested by the user

_DCR between a row and a dataset_
- Given a row `s` and a dataset `D`, this algorithm computes the distance to the closest record of `D`.
- Compute the record distance between `s` and every single row in `D` using the DCR between two rows algorithm below.
- The DCR is the minimum distance value that we compute between `s` and all rows of `D`.

_DCR between 2 rows (including null values)_
Given a row `s` and a row `d` compute distance between these rows.
- Loop through every column's value (j) in these rows and compute the distance. If these values are called `s_j` and `d_j`, then distance between them is defined based on the type of data:
  - For numerical data, `distance(s_j, d_j) = abs(s_j - d_j) / (range of column)` if both `s_j` and `d_j` are both non-null.
    - The `distance(s_j, d_j) = 0` if both `s_j` and `d_j` are null.
    - The `distance(s_j, d_j) = 1` if only one of `s_j` and `d_j` are null.
  - For categorical data, nulls are considered another category and treated like any other category value.
    - `distance(s_j, d_j) = 0 if s_j == d_j else 1`

The overall distance between rows `s` and `d` is the average of the column distances we compute

#### `compute`
The `compute` method should take the same arguments as the `compute_breakdown` method.

The `compute` method should just return the overall `score` parameter calculated by `compute_breakdown`.

### Additional context
See [doc](https://docs.google.com/document/d/1APIEvLfCCptC1bbFgF8bBRLnyAC_3y5pICb_9yCUxuA/edit?tab=t.jkxbunij8p7p#heading=h.rqxtwadpvg5j)

### LINKED ISSUE #739: DCR metrics should allow the possibility of constant numerical columns
### Problem Description
The DCRBaselineProtection and DCROverfittingProtection metrics measure the privacy of synthetic data by measuring how far away it is from the real data. Underlying both of these metrics is a computation that measures the "distance" between 2 rows of data. It is defined in issue #720 (pasted below):

Image

**The problem**: Numerical data may sometimes have a range of 0. This means that in the real data, all values in the column were exactly the same. 

Currently, both DCR metric would crash on such a column because `col_range=0` leads to a `ZeroDivisionError`.

### Expected behavior
If, for numerical data column, the real column range = 0, then compute the distance using the same methodology we do for categorical column. That is to say: `distance(s_j, d_j) = 0` if `s_j == d_j` else the distance is `1`

### PR DESCRIPTION
**Pull request by lajohn4747**
This is the feature branch for #719 and #720.

resolves #719 
resolves #720 
resolves #739 
CU-86b3w869f
CU-86b3w7pbq

This branch consists of only approved PRs:
#726 
#728 
#733 
#745 
#752 
#746 
#751 
#754 
#756 
#760

### CODE REVIEW

**Comment by sdv-team** [comment]
Task linked: [CU-86b3w7pbq SDMetrics - Add `DCROverfittingProtection` metric #719](https://app.clickup.com/t/86b3w7pbq)
Task linked: [CU-86b3w869f SDMetrics - Add `DCRBaselineProtection` metric #720](https://app.clickup.com/t/86b3w869f)

**Comment by lajohn4747** [comment]
> ## [Codecov](https://app.codecov.io/gh/sdv-dev/SDMetrics/pull/735?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sdv-dev) Report
> All modified and coverable lines are covered by tests ✅
> 
> > Project coverage is 83.69%. Comparing base [(`7f08c1e`)](https://app.codecov.io/gh/sdv-dev/SDMetrics/commit/7f08c1e132db1ae977248ace4a9f1633e1ecff08?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sdv-dev) to head [(`5981756`)](https://app.codecov.io/gh/sdv-dev/SDMetrics/commit/5981756d11ecf731a46ddac1664f9a872cfe003a?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sdv-dev).
> 
> > ❗  There is a different number of reports uploaded between BASE ([7f08c1e](https://github.com/sdv-dev/SDMetrics/commit/7f08c1e132db1ae977248ace4a9f1633e1ecff08)) and HEAD ([5981756](https://github.com/sdv-dev/SDMetrics/commit/5981756d11ecf731a46ddac1664f9a872cfe003a)). Click for more details.
> > HEAD has 1 upload less than BASE
> > 
> > 
> > 
> > Flag
> > BASE ([7f08c1e](https://github.com/sdv-dev/SDMetrics/commit/7f08c1e132db1ae977248ace4a9f1633e1ecff08))
> > HEAD ([5981756](https://github.com/sdv-dev/SDMetrics/commit/5981756d11ecf731a46ddac1664f9a872cfe003a))
> > 
> > 
> > 
> > 
> > integration
> > 1
> > 0
> 
> Additional details and impacted files
> ```diff
> @@             Coverage Diff             @@
> ##             main     #735       +/-   ##
> ===========================================
> - Coverage   95.29%   83.69%   -11.60%     
> ===========================================
>   Files         112      115        +3     
>   Lines        4378     4570      +192     
> ===========================================
> - Hits         4172     3825      -347     
> - Misses        206      745      +539     
> ```
> 
> [Flag](https://app.codecov.io/gh/sdv-dev/SDMetrics/pull/735/flags?src=pr&el=flags&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sdv-dev)	Coverage Δ	
> [integration](https://app.codecov.io/gh/sdv-dev/SDMetrics/pull/735/flags?src=pr&el=flag&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sdv-dev)	`?`	
> [unit](https://app.codecov.io/gh/sdv-dev/SDMetrics/pull/735/flags?src=pr&el=flag&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sdv-dev)	`83.69% <100.00%> (+0.71%)`	⬆️
> Flags with carried forward coverage won't be shown. [Click here](https://docs.codecov.io/docs/carryforward-flags?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sdv-dev#carryforward-flags-in-the-pull-request-comment) to find out more.
> 
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/sdv-dev/SDMetrics/pull/735?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sdv-dev). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=sdv-dev).

Strange as each of these PRs, code coverage was improved as it tested everything but the combination seems to throw it off.

**Review by R-Palazzo — COMMENTED** [review]
Thanks, @lajohn4747!
Since this is a feature branch, could you try rebasing it from main using
`git rebase -i main`?
This way, the branch will be up to date with `main`, and we will mostly maintain one commit per issue solved, which is our goal for any feature branch.
Also because we use `rebase and merge` for the feature branch rather than the usual `squash and merge` for other PRs.

**Review by amontanez24 — APPROVED** [review]
